File: python/01/second.py
def solve(input):
    out = 0
    left, right = zip(*input)
    count = {}
    for r in right:
        if r not in count:
            count[r] = 0
        count[r] += 1
    for l in left:
        if l in count:
            out += l * count[l]
    return out


def read(filename):
    with open(filename, "r") as f:
        rows = [row.rstrip() for row in f.readlines()]
        input = []
        for row in rows:
            row = row.split(" ")
            row = map(str.strip, row)
            row = filter(lambda s: s != "", row)
            row = map(int, row)

            input.append(row)
            # input.append(int(row))
            # input.append(list(row))
        return input


def main(filename):
    return solve(read(filename))


if __name__ == "__main__":
    import sys

    name = "input.txt"
    # name = "test.txt"

    if len(sys.argv) < 2:
        print("{}\n".format(main(name)))
    else:
        for f in sys.argv[1:]:
            print("{}:\n{}\n".format(f, main(f)))

File: rust/second.rs
use std::fs;
use std::collections::HashMap;

fn parse_input(contents: String) -> Vec<(i32, i32)> {
    let mut output = Vec::new();
    for row in contents.lines() {
        let mut parts = row.split_whitespace();
        let left = parts.next().unwrap().parse::<i32>().unwrap();
        let right = parts.next().unwrap().parse::<i32>().unwrap();
        output.push((left, right));
    }
    return output;
}

fn solve(rows: Vec<(i32, i32)>) -> i32 {
    let mut output = 0;
    let mut map = HashMap::new();
    for row in rows.iter() {
        let (_l, r) = row;
        let value = map.entry(r).or_insert(0);
        *value += 1;
    }
    for row in rows.iter() {
        let (l, _r) = row;
        if map.contains_key(&l) {
            output += l * map.get(&l).unwrap();
        }
    }
    return output;
}

fn main() {
    let file_name = "input.txt";
    // let file_name = "test.txt";
    let contents = fs::read_to_string(file_name).expect("Should have been able to read the file");
    let rows = parse_input(contents);
    let answer = solve(rows);
    println!("{answer}");
}